### PR TITLE
Clean-up after logger connector redesign 2/2

### LINKED ISSUE #6203: Flush logger connector metrics on Trainer state change
## 🚀 Feature

See title
The logger connector metrics should be grouped by RunningStage.

To be done after #4945

This will facilitate progress bar printing and clean out the code

### Context

https://github.com/PyTorchLightning/pytorch-lightning/blob/110e49dc99194ec4465590775983bce2d4acd4f9/pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py#L349-L352

### LINKED ISSUE #7183: LoggerConnector Refactor
### Pitch

The LoggerConnector Logic is pretty opaque and hard to follow.
The EpochResultStore and HookResult add an extra layer of complexity and the tests are possibly too sparse to catch wrong behaviours.

One of the reason of the complexity is the non-uniformity of the stored logged data. 

Description of internal functionalities:

1. An EpochResultStore is create for each Trainer RUNNING_STAGE.
2 . A new Result Object is created when running a new hook.
     Result Object are enhanced dictionary containing a mapping key - value with the extra logged meta data and inferred      batch_size.
3. Store this Result Object in the associated EpochResultStore. 
     How Result Object are stored is different between TRAIN and TEST/VALIDATION making the code complex and hard to follow.
4. On batch_end: Get the latest stored Result Object and provide its values to logger and progress_bar based on meta.
5. On epoch_end: Reduce the values and provide them to logger and progress_bar based on meta.
    As Logged value can either be a Metric or a float/tensor creating extra internal check for properly reduce on EpochEnd.

Proposition: Uniformize Logged Values to simplify storing them and reduction.

TODOs:
 - [ ] Simplify Result Object internally
 - [ ] Create 1 Result Object for the entire loop.
 - [ ] Storage: RunningStage -> hook_name -> dataloader_idx -> LoggedMetric
 - [ ] Create a LoggedTensorMetric

Here is the pseudo code for the `LoggedMetric`. It will wrap both Metric + tensors and greatly simplify the internal way to store information.
It would also make fault tolerant training simpler as the state could be reduced and stored/reloaded as 'weighted_mean, sum(batch_sizes)' 

```python
import torchmetrics

class LoggedMetric(torchmetrics.Metrics):

      def __init__(self, key, meta, wrap_metric: bool = False):
             self.key = key
             self.meta = meta
             self.wrap_metric = wrap_metric
             if not self.wrap_metric:
                self.add_state("value", default=torch.tensor(0))
                self.add_state("batch_sizes", defauft=[])

      def update(self, value, batch_size):
        if not self.wrap_metric:
            self.value += value
            self.batch_sizes.append(batch_size)
        else:
            if not isinstance(value, torchmetrics.Metrics):
                raise Mis...

            if not hasattr(self, "value"):
                self.value = value
            else:
                if not self.value != value:
                    raise Mis...

    def compute(self):
        if not self.wrap_metric:
            return weighted_mean(self.value, self.batch_sizes)
        else:
            return self.value.compute()

    @property
    def on_epoch(self) -> bool:
        return self.meta["on_epoch"]

    @property
    def on_step(self) -> bool:
        return self.meta["on_step"]

    ...
```

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #7183 
Fixes #6203

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -11,603 +11,534 @@
...
-                on_step=True,
-                on_epoch=False,
-                reduce_fx=reduce_fx,
-                forked=False,
-                dataloader_idx=dataloader_idx,
-            )
+class ResultMetric(Metric):
 
-            self.__setitem__(step_name, value)
+    DTYPE = torch.float32
```
Comment (justusschock) [inline pytorch_lightning/core/step_result.py:80]: Instead of having this, maybe inherit from `Metric` and `DeviceDtypeModuleMixin`?

This would add `self.dtype` and `self.device` and all the logic to carry changes like converting it to half or something similar.

Comment (tchaton) [reply pytorch_lightning/core/step_result.py]: Done !

File: tests/trainer/loops/test_evaluation_loop_flow.py [hunk 2]
```diff
@@ -65,6 +66,8 @@ def backward(self, loss, optimizer, optimizer_idx):
     assert not model.validation_step_end_called
     assert not model.validation_epoch_end_called
 
+    trainer.state.stage = RunningStage.TRAINING
```
Comment (awaelchli) [inline tests/trainer/loops/test_evaluation_loop_flow.py:69]: this is here to simulate the behavior we are still training right? so that you can make the assertions below?

File: pytorch_lightning/trainer/training_loop.py [hunk 3]
```diff
@@ -83,6 +82,8 @@ def __init__(
         else:
             self.trainer.num_sanity_val_steps = num_sanity_val_steps
 
+        self.train_results = ResultCollection(True)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:85]: are these the training results for one epoch, and after one epoch it will be cleared right?
I can't see where it gets updated?

**Review by awaelchli — COMMENTED** [review]
looks like a lot of great improvements!

some first high level comments:

File: pytorch_lightning/trainer/training_loop.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:85]: Gets updated by reference through the `trainer.result_collection` property

File: pytorch_lightning/trainer/training_loop.py [hunk 4]
```diff
@@ -493,21 +479,20 @@ def run_training_epoch(self):
             if batch_output.signal == -1:
                 break
 
+            # -----------------------------------------
+            # SAVE METRICS TO LOGGERS AND PROGRESS_BAR
+            # -----------------------------------------
+            self.trainer.logger_connector.update_train_step_metrics(batch_output)
```
Comment (carmocca) [inline pytorch_lightning/trainer/training_loop.py:485]: @tchaton why is this moved to above `on_train_batch_end`?

`on_trrain_batch_end` can also log

File: pytorch_lightning/trainer/training_loop.py [hunk 5]
```diff
@@ -670,30 +637,28 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
...
-        return output
 
-    def _run_optimization(self, batch_idx, split_idx, split_batch, opt_idx=0, optimizer=None):
+    def _run_optimization(self, batch_idx, split_batch, opt_idx=0, optimizer=None):
         # TODO: In v1.5, when optimizer_idx gets removed from training_step in manual_optimization, change
         #   opt_idx=0 to opt_idx=None in the signature here
 
         # toggle model params + set info to logger_connector
-        self.run_train_split_start(split_idx, split_batch, opt_idx, optimizer)
+        self.run_train_split_start(batch_idx, split_batch, opt_idx, optimizer)
```
Comment (awaelchli) [inline pytorch_lightning/trainer/training_loop.py:661]: this doesn't look right. accidental rename?

Comment (carmocca) [reply pytorch_lightning/trainer/training_loop.py:661]: @tchaton

Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:658]: Oh yes, there is a bug. It should be `batch_idx * num_splits + split_idx`

File: pytorch_lightning/trainer/training_loop.py (hunk 4 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:470]: My mistake. You are entirely right !

File: pytorch_lightning/trainer/training_loop.py (hunk 5 above)
Comment (tchaton) [reply pytorch_lightning/trainer/training_loop.py:658]: We just need to track the first ever sample seen per epoch to result the metrics.

**Review by awaelchli — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGMT !
